custom_profile_fields: Add "required" parameter to the profile fields.

Vector73 opened the pull request:
<!-- Describe your pull request here.-->
This PR adds a "Required" parameter to the custom profile fields which makes the fields required to be filled by all the members in the organization.

Fixes: #28512<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![Screenshot 2024-02-15 164015](https://github.com/zulip/zulip/assets/119798962/3f256ad4-0a52-4583-97f9-bbba3d9e4343)|![Screenshot 2024-02-15 164028](https://github.com/zulip/zulip/assets/119798962/9a34897a-4238-47d9-abfe-3a29c7c2b900)

![Screenshot 2024-02-14 182057](https://github.com/zulip/zulip/assets/119798962/44cee972-f304-4c37-82d7-4e5fc54ed3c4)
![Screenshot 2024-03-09 181751](https://github.com/zulip/zulip/assets/119798962/b62fc32e-0373-4306-a46d-23d17327069c)

| Dark | Light |
|--------|--------|
|![Screenshot 2024-03-14 165238](https://github.com/zulip/zulip/assets/119798962/a000e764-9eb3-44bc-a673-13d133422252)|![Screenshot 2024-03-14 165255](https://github.com/zulip/zulip/assets/119798962/9a977378-66f1-4e6f-999d-91b39a645021)| 

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Vector73 commented:
I have added a red asterisk in front of the required fields' name in the "Edit profile" form to distinguish them from other fields. This change was not mentioned in the issue description but I think it would be good to differentiate required fields. @alya thoughts?

alya commented:
Thanks! There are definitely problems with the banner implementation, both the colors, and how it affects the sidebars.

Please check your implementation more carefully, and post screenshots in both light and dark theme.

<img width="1453" alt="Screenshot 2024-02-14 at 11 43 42 PM" src="https://github.com/zulip/zulip/assets/2090066/956a09e4-418e-4c48-a7ee-eaad2575eef5">

For the colors, you can use the styling from the desktop notifications top banner for now. But @terpimost FYI about this being an upcoming use case for top banners.

alya commented:
For the field edit/creation modal, I think "Required field" (as specified in the issue) is a better label.

alya commented:
Indicating which fields are required in the profile editing UI is a great idea -- thanks for flagging that!

The red stars feel like they might be a bit much to me. How about we try using regular text color, but also add a tooltip when hovering anywhere over the label for the required profile fields:

> This profile field is required.

Vector73 commented:
> The sidebars sometimes look fine, so I'm not sure what exactly the reproducer is for the state I got in, where you couldn't see 
> the top of either sidebar.

I have fixed this issue. It was occurring when banner was not present initially and was appearing after any required field was made empty. 

Changed the color of the banner and label to "Required field".

Added the tooltip to the required fields' label.

Vector73 commented:
I have tweaked the color of the asterisk. @alya can you please take a look?

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,13 @@ format used by the Zulip server that they are interacting with.
+* [`POST /realm/profile_fields`](/api/create-custom-profile-field),
+[`GET /realm/profile_fields`](/api/get-custom-profile-fields): Added a
+new parameter `required`, which clients use to make field required to
+be filled by all the users in the organization.
```
I'm pretty sure the `register` and `events` endpoints were updated as well, right?

Vector73 replied:
Yeah, I have added them now in "changelog.md".

timabbott commented on web/src/navbar_alerts.js:
```diff
@@ -173,6 +193,8 @@ export function initialize() {
         });
     }
 
+    maybe_show_empty_required_profile_fields_alert();
```
Look more carefully at the construction of `initialize`; it's designed to never show more than one at once.

This new one can be last, after `check_profile_incomplete`

Vector73 replied:
But shouldn't the warning banner still be visible even if there is an info banner in the navbar?

timabbott replied:
The navbar alerts system isn't designed to handle more than one alert at a time; various CSS things I expect to break. So they have an order of precedence; a normal user should never have more than 1 applicable at a time anyway.

Vector73 replied:
OK, I have changed the code accordingly.

Vector73 replied:
OK, I have changed the code accordingly.

timabbott commented on web/src/settings_profile_fields.js:
```diff
@@ -599,6 +603,31 @@ function toggle_display_in_profile_summary_profile_field(e) {
+        field_data,
+        display_in_profile_summary: field.display_in_profile_summary,
+        required: !field.required,
+    };
```
Do you need to pass all these fields, or could it just pass the one field we're changing?

timabbott replied:
It appears the answer is yes; kinda a weird quirk in how the custom profile fields API works compared to the rest of the Zulip API, but I don't think we need to try to change that as part of this PR.

timabbott commented on web/src/state_data.ts:
```diff
@@ -42,6 +42,7 @@ export const realm_schema = z.object({
             id: z.number(),
             name: z.string(),
             order: z.number(),
+            required: z.optional(z.boolean()),
```
Why is this option? It seems like it should just be a `boolean`.

Vector73 replied:
Fixed in the recent commit.

timabbott commented on web/styles/zulip.css:
```diff
@@ -297,10 +297,9 @@ p.n-margin {
+            right: -18px;
             top: 50%;
-            transform: translateY(-50%);
+            transform: translateY(-100%);
```
So this CSS is shared with all the other navbar alerts; in general it is not safe to randomly change such things as part of a commit adding a feature, since it could break the other alerts. If you need this change, can you extract it to a separate commit and provide a commit message explaining why it is necessary and correct?

timabbott replied:
How did you address this question?

Vector73 replied:
This CSS change was required in case of more than one alert. Now that more than one alert is not possible, I have removed this change.

timabbott commented on web/templates/navbar_alerts/empty_required_profile_fields.hbs:
```diff
@@ -0,0 +1,9 @@
+        Your profile is missing the following required fields:
+        {{#each empty_required_profile_fields}}
+            {{this}}{{#if @last}}. {{else}}, {{/if}}
+        {{/each}}
```
These strings, like every other user-visible string in an internationalized web application, needs to be tagged for translation. It seems likely that you'll want to make use of `Intl.ListFormat` to combine together the strings rather than doing this `, ` construction, as well.

See https://zulip.readthedocs.io/en/latest/translating/internationalization.html for important background information on internationalization in software projects and Zulip in particular.

Vector73 replied:
I have fixed it now.

timabbott commented on web/templates/settings/custom_user_profile_field.hbs:
```diff
@@ -1,5 +1,6 @@
 <div class="custom_user_field" name="{{ field.name }}" data-field-id="{{ field.id }}">
-    <label class="inline-block" for="{{ field.name }}" class="title">{{ field.name }}</label>
+    <label class="inline-block {{#if field.required}}required_field{{/if}}" for="{{ field.name }}" class="title">{{ field.name }} </label>
+    {{#if field.required}}<span class="required-symbol">*</span>{{/if}}
```
Is the right place to put the space inside the `label` element or inside this `field.required` clause? I'm not sure why you're adding one just before the `</label>` specifically.

timabbott commented on zerver/models/custom_profile_fields.py:
```diff
@@ -168,6 +169,8 @@ def as_dict(self) -> ProfileDataElementBase:
         if self.display_in_profile_summary:
             data_as_dict["display_in_profile_summary"] = True
+        if self.required:
+            data_as_dict["required"] = True
```
This is probably worth an "#api design" conversation -- I'm not entirely sure why we made `display_in_profile_summary` an optional field, but I think it's possible the reasoning for that doesn't apply to this field, and it should just be always present. (I guess this explains why you had that type in the JS).

Vector73 replied:
Fixed in the recent commit.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -10398,6 +10400,13 @@ paths:
                   example: true
+                required:
+                  description: |
+                    Whether clients are required to fill this profile field.
```
The semantics should be spelled out more -- what is responsible for enforcing it, for example? 

In particular, it should be mentioned that because fields can be added as required fields, clients cannot rely on a `required` field actually being set for every user.

Vector73 replied:
I am not entirely sure I understood what needs to be added here. But I have added a paragraph in zulip.yaml:
> As fields can be added as required fields, clients cannot rely on a required
>  field actually set for every user.

Is this OK?

timabbott replied:
I think the following would be good:

> Whether an organization administrator has configured this profile field as required. 
>
> Because the required property is mutable, clients cannot assume that a required custom profile field has a value. The Zulip web application displays a prominent banner to any user who has not set a value for a required field.

timabbott commented:
This is looking close to being possible to integrate; posted a review of all the technical details.

Vector73 commented:
**Updates in the recent changes**:
- Updated changelog.md to cover ```register``` and ```events``` endpoints.
- Used translation strings in ```empty_required_profile_fields.hbs```.
- Updated required parameter to not make it optional.
- Added space before "*" instead of after ```field.name```.

timabbott commented on web/src/navbar_alerts.js:
```diff
@@ -70,6 +72,32 @@ export function should_show_server_upgrade_notification(ls) {
+            "long",
+            "conjunction",
+        );
+        empty_required_profile_fields_string += ". ";
```
This sort of string manipulation is quite strange and probably not correct internationalization. Why are you doing it?

Vector73 replied:
I did that to add a period at the end of the list. Now I have added it in the "empty_profile_required_fields.hbs" itself:
```handlebars
{{ empty_required_profile_fields }}{{t '.'}}&nbsp;
```

Vector73 replied:
The current implementation adds a tooltip and '*' in front of the required profile fields but there is nothing to show which fields are empty explicitly. We can maybe change the font color of empty required profile fields or add a tip at the top of the edit profile form.

What do you think?

Vector73 replied:
OK, I have added a yellow border around empty required fields. You can take a look in the PR description.

Does this look good to you?

alya replied:
I think red would be better. The color we use in the compose box when someone tries to send an empty message should be fine.

Vector73 replied:
OK, I have changed the border-color to red.

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -473,6 +474,7 @@ export function dispatch_normal_event(event) {
                     if (people.is_valid_bot_user(event.person.user_id)) {
                         settings_users.update_bot_data(event.person.user_id);
                     }
+                    navbar_alerts.maybe_show_empty_required_profile_fields_alert();
```
This should probably be placed within this block within `user_events.update_person`. There's no reason to call that function except for custom profile field updates.

```
    if (Object.hasOwn(person, "custom_profile_field")) {                                                
        people.set_custom_profile_field_data(person.user_id, person.custom_profile_field);              
    }                                                                                                   
```

Vector73 replied:
Fixed in the recent commit.

timabbott commented on web/templates/settings/admin_profile_field_list.hbs:
```diff
@@ -23,6 +23,14 @@
+    <td class="required_cell">
+        <span class="profile_field_required">
+            <label class="checkbox" for="profile_field_required_{{id}}">
+                <input class="required required_checkbox_{{required}}" type="checkbox" id="profile_field_required_{{id}}" {{#if required}} checked="checked" {{/if}} data-profile-field-id="{{id}}"/>
```
`required` might be better called `required-field-toggle` to be more globally unique.

Also we've been generally using `-`s rather than `_`s` as delimiters in new code. See https://zulip.readthedocs.io/en/latest/subsystems/html-css.html#use-clear-unique-names-for-classes-and-object-ids.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,14 @@ format used by the Zulip server that they are interacting with.
+  [`POST /realm/profile_fields`](/api/create-custom-profile-field),
+  [`GET /realm/profile_fields`](/api/get-custom-profile-fields): Added a
+  new parameter `required`, which clients use to make field required to
+  be filled by all the users in the organization.
```
This also is imprecise: 

> Added a new parameter `required` on custom profile field objects, indicating whether an organization administrator has configured the field as something users are required to provide.

Vector73 replied:
Fixed in the recent commit.

Vector73 commented:
**Updates in the recent commit:**
- Fixed navbar alerts so that at most one alert is present in the navbar.
- Updated "changelog.md" and "zulip.yaml" to document changes in more detail.
- Changed ```_```'s to ```-```'s in handlebars files.

alya commented:
Something seems to be off with the checkbox alignment:

![Screenshot 2024-03-13 at 12 17 35 PM](https://github.com/zulip/zulip/assets/2090066/fb8dc8a6-b2cb-47c4-8fc8-f075b565b433)

alya commented:
A few more tweaks:
1. Let's have the red `*` appear only when a required field is empty, and not if there's already content there.
2. The hover target for the tooltip about the field being mandatory should include the `*`.
3. Once you've filled out a mandatory field, we shouldn't let you make it empty again. If it's non-trivial, this can be a follow-up after this PR is merged.

Vector73 commented:
I have addressed (1) and (2) and I think it would be better to do (3) in a separate PR after merging this one.

alya commented:
Hm, the border doesn't look right in light theme. Is the PR somehow adding both a red border and a black one?

![Screenshot 2024-03-13 at 11 56 33 PM](https://github.com/zulip/zulip/assets/2090066/488660a3-fcc8-4722-87d5-3b17cb23bf7f)

Please be sure to test any visual changes in both themes. Posting screenshots with both light and dark theme in the PR description will be a good way to remind yourself to do so.

Vector73 commented:
I did update the screenshots in the PR description but didn't notice the black outline behind the red border. I have fixed that now and updated the screenshots.

alya commented:
Cool, thanks, that screenshot looks good to me.

alya commented:
@timabbott Do you want to take another look now?

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -100,11 +100,20 @@ export function dispatch_normal_event(event) {
+                navbar_alerts.maybe_show_empty_required_profile_fields_alert();
+            }
             break;
+        }
```
I would not expect you to need to add these outer braces here; is there a reason for this inconsistency with the existing style?

Vector73 replied:
It was to suppress the linting error when I was defining a variable in case block ```const $navbar_alert = $("#navbar_alerts_wrapper").children(".alert").first();```. It is not needed now though.

timabbott replied:
I see. In this file in particular, generally the goal is to have modules responsible for specific subsystems do anything complicated, just because otherwise this file would be infinitely long anyway.

timabbott commented on web/src/user_events.js:
```diff
@@ -128,6 +129,38 @@ export const update_person = function update(person) {
+            $navbar_alert.is(":hidden")
+        ) {
+            navbar_alerts.maybe_show_empty_required_profile_fields_alert();
+        }
```
This block is repeated in a few places; I think it should probably be a function exported by the `navbar_alerts` module. 

Or maybe just part of the `navbar_alerts.maybe_show_empty_required_profile_fields_alert` interface? 

In any case, you probably also only need to do this check if `person.user_id == current_user.user_id`, and that seems like an appropriate outer conditional.

Vector73 replied:
I have transferred that block in ```navbar_alerts.maybe_show_empty_required_profile_fields_alert``` now.

timabbott commented on web/src/user_events.js:
```diff
@@ -128,6 +129,38 @@ export const update_person = function update(person) {
+                    $required_symbol.addClass("hidden");
+                }
+            }
+        }
```
The rest of this logic also seems to assume that the user whose data was edited is the current user. This is the kind of bug that's very hard to catch while testing (though testing live-update with another user account is highly recommended and might catch this) -- you really need to always be thinking when writing logic "what is the exact condition for when this should happen?".

Vector73 replied:
I thought this block of code is actually needed even if the fields that are updated are of another user in the "Manage user" tab of the user profile modal, but the user profile modal closes after changes are saved. So it is redundant for that use case right now, but there is an open issue ([#26692](https://github.com/zulip/zulip/issues/26692)) that might change this in the future. I am putting it in ```person.user_id === people.my_current_user_id()``` condition for now though.

timabbott replied:
Just to be concrete, is this the case where you as an administrator are editing a "required" field for another user via their user profile?

I guess I'm not sure how the styling for that should work -- does this colored outline apply in that case? I guess that's reasonable, but I think even once we do #26692, I'd expect us to want to handle these updates by just rerendering the profile.

But what this discussion does cause me to realize is that your selector, `.custom_user_field[data-field-id="${CSS.escape(field_id)}"]`, is insufficiently precise -- will that change the styling on __someone else's__ profile if one of your custom profile fields is edited in another window?

Vector73 replied:
Oh, you are right. I missed that. I have fixed it now.

timabbott commented:
I posted another batch of comments; other than those, this looks good to me, though I expect I'll want to do a round of manual testing after those changes before merging.

timabbott commented on web/src/navbar_alerts.js:
```diff
@@ -70,6 +72,28 @@ export function should_show_server_upgrade_notification(ls) {
+        !$navbar_alert?.length ||
+        $navbar_alert.is("#empty-required-profile-fields-warning") ||
+        $navbar_alert.is(":hidden")
+    ) {
```
Can you talk about why this is the correct check? I don't really see why you're looking at `:hidden`, for example.

Vector73 replied:
This is used to check if there are any other alerts present in the navbar. When we close the already present alerts they get hidden. So ```:hidden``` check is for adding the warning banner when already present alerts are closed. The ```$navbar_alert.is("#empty-required-profile-fields-warning")``` condition is used to remove the warning banner when the user fills all the required fields.

timabbott replied:
OK. This seems to pretty weirdly misuse the APIs for this module. I made the following changes, which seem to produce a much cleaner result with less code duplication:

``` diff
diff --git a/web/src/navbar_alerts.js b/web/src/navbar_alerts.js
index f47ac04fcb..13cb67ea68 100644
--- a/web/src/navbar_alerts.js
+++ b/web/src/navbar_alerts.js
@@ -73,24 +73,26 @@ export function should_show_server_upgrade_notification(ls) {
 }
 
 export function maybe_show_empty_required_profile_fields_alert() {
+    const empty_required_profile_fields_exist = realm.custom_profile_fields
+          .map((f) => ({
+              ...f,
+              value: people.my_custom_profile_data(f.id)?.value,
+          }))
+          .find((f) => f.required && !f.value);
+    if (!empty_required_profile_fields_exist) {
+        return;
+    }
+
     const $navbar_alert = $("#navbar_alerts_wrapper").children(".alert").first();
     if (
         !$navbar_alert?.length ||
         $navbar_alert.is("#empty-required-profile-fields-warning") ||
         $navbar_alert.is(":hidden")
     ) {
-        const empty_required_profile_fields_exist = realm.custom_profile_fields
-            .map((f) => ({
-                ...f,
-                value: people.my_custom_profile_data(f.id)?.value,
-            }))
-            .find((f) => f.required && !f.value);
-
-        $("#empty-required-profile-fields-warning").remove();
-        if (empty_required_profile_fields_exist) {
-            $("#navbar_alerts_wrapper").html(render_empty_required_profile_fields());
-        }
-        $(window).trigger("resize");
+        open({
+            data_process: "profile-missing-required",
+            rendered_alert_content_html: render_empty_required_profile_fields(),
+        });
     }
 }
 
diff --git a/web/templates/navbar_alerts/empty_required_profile_fields.hbs b/web/templates/navbar_alerts/empty_required_profile_fields.hbs
index 8c075fcdcb..a74eafa37e 100644
--- a/web/templates/navbar_alerts/empty_required_profile_fields.hbs
+++ b/web/templates/navbar_alerts/empty_required_profile_fields.hbs
@@ -1,4 +1,4 @@
-<div id="empty-required-profile-fields-warning" class="alert alert-info">
+<div data-step="1">
     <span>
         {{t 'Your profile is missing required fields.'}}&nbsp;
         <a class="alert-link" href="#settings/profile">{{t 'Edit your profile'}}</a>
```

timabbott commented:
I think we're down to just a couple more comments, thanks for the revisions @Vector73.

timabbott commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -527,6 +539,7 @@ def test_update(self) -> None:
         self.assertEqual(field.hint, "New contact number")
         self.assertEqual(field.field_type, CustomProfileField.SHORT_TEXT)
         self.assertEqual(field.display_in_profile_summary, True)
+        self.assertEqual(field.required, True)
```
I'm in the process of merging this; adding an assert that this is `False` earlier in the test; otherwise, you've not verified that you actually changed it.

See https://zulip.readthedocs.io/en/latest/testing/philosophy.html for some other helpful advice on writing good tests.

timabbott commented on web/styles/settings.css:
```diff
@@ -672,6 +674,15 @@ input[type="checkbox"] {
+.field {
+    margin-top: 5px;
+    max-width: fit-content;
+}
```
This sort of refactoring to create an extremely generic `field` CSS rule is very dangerous. The following grep shows how this will modify how bot avatar rows, the change password form, etc. are rendered, which seems not the intent.
``` console
$ git grep class= | grep [^-_]field[^-_]
templates/corporate/for/business.html:                    <li><div class="list-content">Use Zulip as your employee directory. Sync names, <a href="/help/roles-and-permissions">roles</a>, avatars, and <a href="/help/custom-profile-fields">custom profile fields</a> from LDAP or SAML.</div></li>
templates/corporate/support/support_request.html:            <fieldset class="support-request">
templates/zerver/register.html:            <fieldset class="org-registration">
templates/zerver/register.html:            <fieldset class="user-registration">
web/templates/dialog_change_password.hbs:    <div class="field password-div">
web/templates/dialog_change_password.hbs:    <div class="field password-div">
web/templates/settings/admin_profile_field_list.hbs:        <button class="button rounded small btn-warning open-edit-form-modal tippy-zulip-delayed-tooltip" data-tippy-content="{{t 'Edit custom profile field' }}" data-profile-field-id="{{id}}">
web/templates/settings/admin_profile_field_list.hbs:        <button class="button rounded small delete btn-danger tippy-zulip-delayed-tooltip" data-tippy-content="{{t 'Delete custom profile field' }}" data-profile-field-id="{{id}}">
web/templates/settings/bot_avatar_row.hbs:            <div class="field">{{t "Bot type" }}</div>
web/templates/settings/bot_avatar_row.hbs:            <div class="field">{{t "Bot email" }}</div>
web/templates/settings/bot_avatar_row.hbs:            <span class="field">{{t "API key" }}</span>
web/templates/settings/custom_user_profile_field.hbs:<div class="custom_user_field" name="{{ field.name }}" data-field-id="{{ field.id }}">
web/templates/settings/custom_user_profile_field.hbs:        <label class="inline-block {{#if field.required}}required-field{{/if}}" for="{{ field.name }}" class="title">{{ field.name }}</label>
web/templates/settings/custom_user_profile_field.hbs:    <div class="field_hint">{{ field.hint }}</div>
web/templates/settings/custom_user_profile_field.hbs:    <div class="field {{#if is_empty_required_field}}empty-required-field{{/if}}">
web/templates/settings/custom_user_profile_field.hbs:        <input class="custom_user_field_value datepicker {{#if for_manage_user_modal}}modal_text_input{{else}}settings_text_input{{/if}}" data-field-id="{{ field.id }}" type="text"
web/templates/settings/profile_field_settings_admin.hbs:        <button class="button rounded sea-green" id="add-custom-profile-field-btn">{{t "Add a new profile field" }}</button>
web/templates/settings/profile_field_settings_admin.hbs:            <tbody id="admin_profile_fields_table" class="required-text" data-empty="{{t 'No custom profile field configured for this organization.' }}"></tbody>
web/templates/settings_overlay.hbs:                        <div class="text">{{t "Custom profile fields" }}</div>
````

Never do this sort of refactor that could impact random other parts of the app inside a giant commit adding a feature; it's a very very easy way to break unexpected things, and takes a lot of attention from a reviewer to notice you've done it. You should always do it in its own commit, which draws attention to the change, both for yourself and for reviewers, and forces you to think about writing a commit message for how you can provide it's correct ... or whether it's in fact not.

timabbott replied:
I'm adding three little prep commits removing the other users of `.field` as a class name, and also renamed `.field` here as well.

@karlstolley FYI -- these commits touch very lightly on the settings CSS changes you're doing, so you might get some logical or otherwise merge commits with your `dropdown-title` work.

timabbott commented:
OK I pushed back to this PR after a bunch of changes, noted above. Thanks for making this happen @Vector73!

I hope you find the feedback helpful! This is definitely an improvement from some of your previous PRs, but I think you still need to focus on your attention to detail in implementation work; hopefully some of the comments above will help give you a sense for what types of details to focus on.

Vector73 commented:
Thanks for the feedback @timabbott. I will keep that in mind in the future and be more conscious about the changes that I am implementing.

timabbott commented:
Fixed a linter error; this should merge now.
